### PR TITLE
Fix panic on ARM, 386, and 32-bit MIPS by aligne fileds used by atomic package

### PR DESCRIPTION
**Pull request by xoac**
When I tried to use my program on ARM I got panic:
```
(...)
panic: unaligned 64-bit atomic operation [recovered]
        panic: unaligned 64-bit atomic operation

goroutine 1 [running]:
github.com/alecthomas/kong.catch(0x2cb1f58)
        /home/sr/go/pkg/mod/github.com/alecthomas/kong@v0.2.18-0.20210713035501-d1a818b5a1a8/kong.go:407 +0xa8
panic(0x6b9540, 0x8b0528)
        /usr/lib/go/src/runtime/panic.go:965 +0x174
runtime/internal/atomic.panicUnaligned()
        /usr/lib/go/src/runtime/internal/atomic/unaligned.go:8 +0x24
runtime/internal/atomic.Xadd64(0x2c77b14, 0x1, 0x0, 0x7, 0xc)
        /usr/lib/go/src/runtime/internal/atomic/asm_arm.s:233 +0x14
go.mongodb.org/mongo-driver/x/mongo/driver/topology.(*pool).makeNewConnection(0x2c77ae0, 0x2d62c01, 0x0, 0x1, 0x0, 0x0)
        /home/sr/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.1/x/mongo/driver/topology/pool.go:287 +0x94
go.mongodb.org/mongo-driver/x/mongo/driver/topology.(*pool).get(0x2c77ae0, 0x8bc73c, 0x2cf99c0, 0x0, 0x0, 0x2cc8988)
        /home/sr/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.1/x/mongo/driver/topology/pool.go:440 +0x3b0
go.mongodb.org/mongo-driver/x/mongo/driver/topology.(*Server).Connection(0x2d96f30, 0x8bc73c, 0x2cf99c0, 0x0, 0x0, 0x0, 0x0)
        /home/sr/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.1/x/mongo/driver/topology/server.go:268 +0x5c
go.mongodb.org/mongo-driver/x/mongo/driver.Operation.getServerAndConnection(0x2dae628, 0x752d82, 0xb, 0x8ba99c, 0x2d87580, 0x2dae630, 0x8b84d4, 0x2db3680, 0x0, 0x0, ...)
        /home/sr/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.1/x/mongo/driver/operation.go:246 +0x94
go.mongodb.org/mongo-driver/x/mongo/driver.Operation.Execute(0x2dae628, 0x752d82, 0xb, 0x8ba99c, 0x2d87580, 0x2dae630, 0x8b84d4, 0x2db3680, 0x0, 0x0, ...)
        /home/sr/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.1/x/mongo/driver/operation.go:301 +0x80
go.mongodb.org/mongo-driver/x/mongo/driver/operation.(*Insert).Execute(0x2d9aea0, 0x8bc73c, 0x2cf99c0, 0x2dae620, 0x9f3c141e)
        /home/sr/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.1/x/mongo/driver/operation/insert.go:109 +0x1a4
go.mongodb.org/mongo-driver/mongo.(*Collection).insert(0x2d83440, 0x8bc73c, 0x2cf99c0, 0x2cb1bf4, 0x1, 0x1, 0x2cb1be8, 0x1, 0x1, 0x0, ...)
        /home/sr/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.1/mongo/collection.go:297 +0x54c
go.mongodb.org/mongo-driver/mongo.(*Collection).InsertOne(0x2d83440, 0x8bc73c, 0x2cf99c0, 0x690128, 0x2d9ae40, 0x0, 0x0, 0x0, 0xc0425a05, 0x82bfeb58, ...)
        /home/sr/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.1/mongo/collection.go:336 +0xd4
 (...)
```

The atomic package has a [bug section](https://pkg.go.dev/sync/atomic#pkg-note-BUG) with this note:
> On ARM, 386, and 32-bit MIPS, it is the caller's responsibility to arrange for 64-bit alignment of 64-bit words accessed atomically. The first word in a variable or in an allocated struct, array, or slice can be relied upon to be 64-bit aligned. 

I have tested it with my program after change (on arm target) and it worked.

### CODE REVIEW

File: x/mongo/driver/topology/pool.go
Comment (kevinAlbs) [inline x/mongo/driver/topology/pool.go:69]:
```suggestion
	_alignment64bits            int32 // details of atomic bug: https://pkg.go.dev/sync/atomic#pkg-note-BUG
```

To fix go lint, which logged this warning:

```
2021/08/27 16:09:23.899] golint ./bson ./bson/bsoncodec ./bson/bsonoptions ./bson/bsonrw ./bson/bsonrw/bsonrwtest ./bson/bsontype ./bson/mgocompat ./bson/primitive ./event ./mongo ./mongo/address ./mongo/description ./mongo/gridfs ./mongo/integration ./mongo/integration/mtest ./mongo/integration/unified ./mongo/options ./mongo/readconcern ./mongo/readpref ./mongo/testatlas ./mongo/testaws ./mongo/writeconcern ./x/bsonx ./x/bsonx/bsoncore ./x/mongo/driver ./x/mongo/driver/auth ./x/mongo/driver/auth/internal/awsv4 ./x/mongo/driver/connstring ./x/mongo/driver/dns ./x/mongo/driver/drivergen ./x/mongo/driver/drivertest ./x/mongo/driver/examples/cluster_monitoring ./x/mongo/driver/examples/count ./x/mongo/driver/examples/server_monitoring ./x/mongo/driver/examples/workload ./x/mongo/driver/integration ./x/mongo/driver/mongocrypt ./x/mongo/driver/mongocrypt/options ./x/mongo/driver/ocsp ./x/mongo/driver/operation ./x/mongo/driver/session ./x/mongo/driver/topology ./x/mongo/driver/uuid ./x/mongo/driver/wiremessage ./tag ./examples/documentation_examples | ./etc/lintscreen.pl .lint-allowlist
[2021/08/27 16:09:26.104] x/mongo/driver/topology/pool.go:69:2: don't use underscores in Go names; struct field _alignment_64bits should be _alignment64bits
[2021/08/27 16:09:26.105] Makefile:62: recipe for target 'lint' failed
```

**Review by kevinAlbs — COMMENTED** [review]
@xoac, do you have a reproducible program available? I was unable to reproduce the crash on an Ubuntu 20.04 ARM64 host.

**Comment by xoac** [comment]
> I was unable to reproduce the crash on an Ubuntu 20.04 ARM64 host.

@kevinAlbs That's correct this bug is available ARM (in my case armv7hf) not ARM64 (or any 64-bit platform). To reproduce this bug you need 32-bit platform for example Raspberry or in my case I used beaglbone. 

I think this bug should be reproducible on emulated machine. So maybe https://github.com/multiarch/qemu-user-static could be considered for confirmation.

**Comment by benjirewis** [comment]
Hello again @xoac, and apologies for the delay! We've been trying to reproduce the error, and I was _just_ able to do so using those QEMU images you provided (thanks 😄 ). 

Could you change the type of `connected` to be `Int64` instead of adding the extra `_alignment64bits` field? And, could you move the `connected` field to the top of the `pool` struct? Using `Int64`s as the first field seems to be the [suggested method](https://go101.org/article/memory-layout.html) of aligning well on 32-bit and 64-bit architectures.

Note that changing `connected` to be `Int64` will create some errors throughout the repo. You'll have to rename `LoadInt32` and `StoreInt32` to `LoadInt64` and `StoreInt64`, respectively. You'll also have to change the type of `topology.connection.connected` to be `Int64`; could you move that `connected` field to the top of the `connection` struct?

Thanks so much for your contribution.

**Comment by xoac** [comment]
I have made changes according to your suggestion (thanks) and waiting for feedback. Maybe it would be good idea to add testing with QEMU in CI or adding tests with `unsafe.Alignof()` but I am not familiar with this function at this moment.

File: x/mongo/driver/topology/connection.go
Comment (benjirewis) [inline x/mongo/driver/topology/connection.go:42]:
```suggestion
	// connected must be accessed using the atomic package and should be at the beginning of the struct.
```

File: x/mongo/driver/topology/pool.go
Comment (benjirewis) [inline x/mongo/driver/topology/pool.go:61]:
```suggestion
	// These fields must be accessed using the atomic package and should be at the beginning of the struct.
```

File: x/mongo/driver/topology/pool_generation_counter.go
Comment (benjirewis) [inline x/mongo/driver/topology/pool_generation_counter.go:27]:
```suggestion
	// state must be accessed using the atomic package and should be at the beginning of the struct.
```

File: x/mongo/driver/topology/topology_errors_test.go
Comment (benjirewis) [inline x/mongo/driver/topology/topology_errors_test.go:7]: Is this necessary?

**Review by benjirewis — CHANGES_REQUESTED** [review]
Perfect! Thanks for the quick turnaround @xoac. Just a few small suggestions and a question. 

We're planning on adding the [`atomicalign`](https://pkg.go.dev/golang.org/x/tools/go/analysis/passes/atomicalign) go-vet flag to our CI as part of upcoming work in GODRIVER-1066. That should detect bugs like this when building.

File: x/mongo/driver/topology/topology_errors_test.go
Comment (xoac) [reply x/mongo/driver/topology/topology_errors_test.go]: It's not, but start with go 1.17 this is suggested way to add build tags. So I think there should be a period when `//go:build` and `// +build` are used together and then `// +build` is removed. In my case `//go:build` was automatically added on save by vsc extension that I disable to remove this line.

**Review by benjirewis — APPROVED** [review]
LGTM pending tests passing. Thanks so much for your contribution @xoac 😄

**Review by kevinAlbs — APPROVED** [review]
LGTM. Thanks again for the contribution @xoac!

**Comment by benjirewis** [comment]
Backported to `release/1.7` with [this commit](https://github.com/mongodb/mongo-go-driver/commit/977c9aeb4c80d44951f68c155eb9d69904216bfe).

**Comment by FBlass** [comment]
I still get this issue on a Raspberry Pi 2 Model B Rev 1.1 (ARMv7 Processor rev 5 (v7l)) cross compiled from windows with 
GOOS=linux
GOARCH=arm
GOARM=6

```
panic: unaligned 64-bit atomic operation  
                                                                                                                                                                                                                                                                                                                                                          
goroutine 1 [running]:                                                                                                                                                                            
runtime/internal/atomic.panicUnaligned()                                                                                                                                                                  
		C:/Program Files/Go/src/runtime/internal/atomic/unaligned.go:8 +0x24
runtime/internal/atomic.Cas64(0x252814c, 0x0, 0x2)                                                                                                                                                        
		C:/Program Files/Go/src/runtime/internal/atomic/atomic_arm.s:241 +0x14
go.mongodb.org/mongo-driver/x/mongo/driver/topology.(*Server).Connect(0x2528140, 0x240ef68)
        C:/Users/x/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.4/x/mongo/driver/topology/server.go:198 +0x44
go.mongodb.org/mongo-driver/x/mongo/driver/topology.ConnectServer({0x2420690, 0x2c}, 0x240ef68, {0x61, 0x94, 0xcd, 0x16, 0xe2, 0xa4, 0x25, ...}, ...)
        C:/Users/x/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.4/x/mongo/driver/topology/server.go:136 +0x84
go.mongodb.org/mongo-driver/x/mongo/driver/topology.(*Topology).addServer(0x2510300, {0x2420690, 0x2c})
        C:/Users/x/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.4/x/mongo/driver/topology/topology.go:686 +0xb8
go.mongodb.org/mongo-driver/x/mongo/driver/topology.(*Topology).Connect(0x2510300)
        C:/Users/x/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.4/x/mongo/driver/topology/topology.go:213 +0x660
go.mongodb.org/mongo-driver/mongo.(*Client).Connect(0x2594400, {0x625a74, 0x241a084})
        C:/Users/x/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.4/mongo/client.go:156 +0x4c
go.mongodb.org/mongo-driver/mongo.Connect({0x625a74, 0x241a084}, {0x2539eb8, 0x1, 0x1})                                                                                                                   
		C:/Users/x/go/pkg/mod/go.mongodb.org/mongo-driver@v1.7.4/mongo/client.go:106 +0x60                                                                                                    
main.InitDatabase()                                                                                                                                                                                       
		C:/Users/x/go/src/immoScoutCrawler/mongo.go:29 +0x120                                                                                                                                 
main.main()
```

The issue is only completely fixed when moving the [connectionstate in the Server struct](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L87) up.

Looking [into this Thread](https://www.mongodb.com/community/forums/t/cant-connect-to-mongodb-using-golang/126622/4) it seems like I am not the only one having this issue.

**Comment by benjirewis** [comment]
Thanks for following up @FBlass; it looks like `connectionstate` should be at the top of the `topology.Server` struct. Filed GODRIVER-2236.

**Pull request merged**
